src/check/property/AsyncProperty.generic.ts

```diff
@@ -40,9 +41,18 @@ export interface IAsyncPropertyWithHooks<Ts> extends IAsyncProperty<Ts> {
  */
 export class AsyncProperty<Ts> implements IAsyncPropertyWithHooks<Ts> {
   static dummyHook: HookFunction = () => {};
-  private beforeEachHook: HookFunction = AsyncProperty.dummyHook;
-  private afterEachHook: HookFunction = AsyncProperty.dummyHook;
-  constructor(readonly arb: Arbitrary<Ts>, readonly predicate: (t: Ts) => Promise<boolean | void>) {}
+  private readonly globalBeforeEachHook: HookFunction;
+  private readonly globalAfterEachHook: HookFunction;
+  private beforeEachHook: HookFunction;
+  private afterEachHook: HookFunction;
+  constructor(readonly arb: Arbitrary<Ts>, readonly predicate: (t: Ts) => Promise<boolean | void>) {
+    const { asyncBeforeEach = AsyncProperty.dummyHook, asyncAfterEach = AsyncProperty.dummyHook } =
+      readConfigureGlobal() || {};
+    this.globalBeforeEachHook = asyncBeforeEach;
+    this.globalAfterEachHook = asyncAfterEach;
+    this.beforeEachHook = asyncBeforeEach;
+    this.afterEachHook = asyncAfterEach;
```

Comment:
Yeah, I initially implemented it like that. However, I soon realised that that kind of solution wouldn't fit my needs from the issue #876 
e.g. Let's say I have project with tons of property-based tests. Some of them are using beforeEach and/or afterEach. Now I want to add a call to a `cleanup` method globally via one of this new global parameters (e.g. afterEach). With your implementation, I would have to manually add that `cleanup` method to all these properties that were already implementing one/both hook(s).
Quite annoying IMO :sweat_smile:

Reply:
Actually not really, my point is that if you define a value for `beforeEach` (the synchronous one) even asynchronous properties will take them into account.

Reply:
Oh I see what you mean now. So in an example like this:
```typescript
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
```
It should print:
```
global beforeEach
global AfterEach
local afterEach
```
Right?

Reply:
My idea would have been the following:

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
```

Would print:
```
global beforeEach
local afterEach
```
_as global afterEach would have been overriden at call site level_

**AND I need to think about your idea:**

>  it is valuable to not replace the global hooks when new ones have been defined with the beforeEach/afterEach methods

Initially global overrides have been created as a syntaxic sugar to help users define only once a behaviour while it allows to override it locally. If we go for your proposal it means that in the case of beforeEach hooks, override will no be possible. I need to think abour it and come back to you (I will have a look to what Jest or Mocha do on this subject).

Meanwhile if we go in that direction I think that the ordering should be something like : 
- global before
- local before
- local after
- global after

Here are other examples:

<details>
<summary>Global beforeEach, Global afterEach</summary>

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    afterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// global beforeEach
// global afterEach

fc.asyncProperty(() => {});
// global beforeEach
// global afterEach

fc.property(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach
```
</details>

<details>
<summary>Global beforeEach, Global asyncAfterEach</summary>

```js
configureGlobal({
    beforeEach: () => console.log('global beforeEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// OPT1 throws: Async hooks are not supported for synchronous properties. You may override it locally to prevent this error message
// OPT2 throws: Async hooks are not supported for synchronous properties. Please remove your global override or convert it to a synchronous hook or switch to an asynchronous property

fc.asyncProperty(() => {});
// global beforeEach
// global afterEach

fc.property(() => {}).afterEach(() => console.log('local afterEach'));
// OPT1
// global beforeEach
// local afterEach
// OPT2 throws: Async hooks are not supported for synchronous properties. Please remove your global override or convert it to a synchronous hook or switch to an asynchronous property

fc.asyncProperty(() => {}).afterEach(() => console.log('local afterEach'));
// global beforeEach
// local afterEach
```
</details>

<details>
<summary>Global afterEach, Global asyncAfterEach</summary>

```js
configureGlobal({
    afterEach: () => console.log('global afterEach'),
    asyncAfterEach: () => console.log('global afterEach')
});

fc.property(() => {});
// throws: Cannot define both sync and async version of the same hook

fc.asyncProperty(() => {});
// throws: Cannot define both sync and async version of the same hook
```
</details>

Reply:
Concerning your point, what do you think of something like?

```js
fc.asyncProperty(() => {})
  .afterEach((previousAfterEach) => {
    previousAfterEach();
    console.log('local afterEach');
  });
```

or

```js
fc.asyncProperty(() => {})
  .afterEach(() => console.log('local afterEach'), { behaviour: 'keep-existing' });
// behaviour being one of 'keep-existing', 'override' (the default)...
```

or

```js
fc.asyncProperty(() => {})
  .afterEachKeepExisting(() => console.log('local afterEach'));
```

Reply:
Nice! Thanks for thinking about it :slightly_smiling_face: 
I personally find the API of option 1 a bit nicer. However, I think option 2 would give the flexibility of being able to modify the default behaviour via global config:
```typescript
fc.configureGlobal({ defaultBehaviourOfPropertyHooks: 'keep-existing' });
```
But maybe it's over-complicating the API, not sure... I'll leave it to you to decide! :smile:

Reply:
Let's go for option 1! Actually it was my favourite when I wrote the comment and you seem to like it too:

```js
fc.asyncProperty(() => {})
  .afterEach((previousAfterEach) => {
    previousAfterEach();
    console.log('local afterEach');
  });
```